File: orca/whirlpool/python_client/essentials_alpha/whirlpool_essentials/liquiditymath.py
import math
from decimal import Decimal
from .types import TokenAmounts
from .invariant import invariant


class LiquidityMath:
    # https://orca-so.github.io/whirlpools/classes/PoolUtil.html#getTokenAmountsFromLiquidity
    # https://github.com/orca-so/whirlpools/blob/7b9ec35/sdk/src/utils/public/pool-utils.ts#L84
    @staticmethod
    def get_token_amounts_from_liquidity(
        liquidity: int,
        sqrt_price_x64_current: int,
        sqrt_price_x64_lower: int,
        sqrt_price_x64_upper: int,
        round_up: bool
    ) -> TokenAmounts:
        invariant(sqrt_price_x64_lower < sqrt_price_x64_upper, "sqrt_price_x64_lower < sqrt_price_x64_upper")

        liq = Decimal(liquidity)
        lower = Decimal(sqrt_price_x64_lower)
        upper = Decimal(sqrt_price_x64_upper)
        current = min(max(Decimal(sqrt_price_x64_current), lower), upper)  # bounded

        shift_64 = Decimal(2)**64
        token_a = liq * shift_64 * (upper - current) / (current * upper)
        token_b = liq * (current - lower) / shift_64

        if round_up:
            return TokenAmounts(int(math.ceil(token_a)), int(math.ceil(token_b)))
        else:
            return TokenAmounts(int(math.floor(token_a)), int(math.floor(token_b)))

    @staticmethod
    def get_token_a_from_liquidity(
        liquidity: int,
        sqrt_price_x64_current: int,
        sqrt_price_x64_lower: int,
        sqrt_price_x64_upper: int,
        round_up: bool
    ) -> int:
        return LiquidityMath.get_token_amounts_from_liquidity(
            liquidity,
            sqrt_price_x64_current,
            sqrt_price_x64_lower,
            sqrt_price_x64_upper,
            round_up
        ).token_a

    @staticmethod
    def get_token_b_from_liquidity(
            liquidity: int,
            sqrt_price_x64_current: int,
            sqrt_price_x64_lower: int,
            sqrt_price_x64_upper: int,
            round_up: bool
    ) -> int:
        return LiquidityMath.get_token_amounts_from_liquidity(
            liquidity,
            sqrt_price_x64_current,
            sqrt_price_x64_lower,
            sqrt_price_x64_upper,
            round_up
        ).token_b

    # https://github.com/orca-so/whirlpools/blob/7b9ec35/sdk/src/utils/public/pool-utils.ts#L237
    @staticmethod
    def get_liquidity_from_token_a(sqrt_price_x64_0: int, sqrt_price_x64_1: int, amount: int) -> int:
        small_sqrt_price_x64 = min(sqrt_price_x64_0, sqrt_price_x64_1)
        large_sqrt_price_x64 = max(sqrt_price_x64_0, sqrt_price_x64_1)
        if small_sqrt_price_x64 == large_sqrt_price_x64:
            return 0

        # a = L/small_sqrt_price_x64/x64 - L/large_sqrt_price_x64/x64
        # a = L * x64 * ( 1/small_sqrt_price_x64 - 1/large_sqrt_price_x64 )
        # a = L * x64 * ( (large_sqrt_price_x64 - small_sqrt_price_x64) / (small_sqrt_price_x64 * large_sqrt_price_x64) )
        # L = a / x64 / ( (large_sqrt_price_x64 - small_sqrt_price_x64) / (small_sqrt_price_x64 * large_sqrt_price_x64) )
        # L = a * (small_sqrt_price_x64 * large_sqrt_price_x64) / (x64 * (large_sqrt_price_x64 - small_sqrt_price_x64))
        shift_64 = 2**64
        num = amount * (small_sqrt_price_x64 * large_sqrt_price_x64)
        denom = shift_64 * (large_sqrt_price_x64 - small_sqrt_price_x64)
        return num // denom

    # https://github.com/orca-so/whirlpools/blob/7b9ec35/sdk/src/utils/public/pool-utils.ts#L248
    @staticmethod
    def get_liquidity_from_token_b(sqrt_price_x64_0: int, sqrt_price_x64_1: int, amount: int) -> int:
        small_sqrt_price_x64 = min(sqrt_price_x64_0, sqrt_price_x64_1)
        large_sqrt_price_x64 = max(sqrt_price_x64_0, sqrt_price_x64_1)
        if small_sqrt_price_x64 == large_sqrt_price_x64:
            return 0

        # b = L*large_sqrt_price_x64/x64 - L*small_sqrt_sqrt_price_x64/x64
        # L = b / ( large_sqrt_price_x64/x64 - small_sqrt_sqrt_price_x64/x64 )
        # L = b * x64 / ( large_sqrt_price_x64 - small_sqrt_price_x64 )
        shift_64 = 2**64
        num = amount * shift_64
        denom = large_sqrt_price_x64 - small_sqrt_price_x64
        return num // denom

    # https://orca-so.github.io/whirlpools/classes/PoolUtil.html#estimateLiquidityFromTokenAmounts
    # https://github.com/orca-so/whirlpools/blob/7b9ec35/sdk/src/utils/public/pool-utils.ts#L141
    @staticmethod
    def get_max_liquidity_from_token_amounts(
        sqrt_price_x64_current: int,
        sqrt_price_x64_lower: int,
        sqrt_price_x64_upper: int,
        amounts: TokenAmounts
    ) -> int:
        invariant(sqrt_price_x64_lower < sqrt_price_x64_upper, "sqrt_price_x64_lower < sqrt_price_x64_upper")

        if sqrt_price_x64_current >= sqrt_price_x64_upper:
            return LiquidityMath.get_liquidity_from_token_b(sqrt_price_x64_lower, sqrt_price_x64_upper, amounts.token_b)
        elif sqrt_price_x64_current <= sqrt_price_x64_lower:
            return LiquidityMath.get_liquidity_from_token_a(sqrt_price_x64_lower, sqrt_price_x64_upper, amounts.token_a)
        else:
            liquidity_a = LiquidityMath.get_liquidity_from_token_a(sqrt_price_x64_current, sqrt_price_x64_upper, amounts.token_a)
            liquidity_b = LiquidityMath.get_liquidity_from_token_b(sqrt_price_x64_lower, sqrt_price_x64_current, amounts.token_b)
            return min(liquidity_a, liquidity_b)

File: orca/pool/rust_client/with_anchor_client/src/whirlpool_essential/liquidity_math.rs
//use rust_decimal::prelude::*;
//use rust_decimal::MathematicalOps;
use bigdecimal::{ToPrimitive, FromPrimitive, BigDecimal};
use num_bigint::BigInt;

pub struct TokenAmounts {
  pub token_a: u64,
  pub token_b: u64,
}

pub fn get_token_amounts_from_liquidity(
  liquidity: u128,
  sqrt_price_x64_current: u128,
  sqrt_price_x64_lower: u128,
  sqrt_price_x64_upper: u128,
  round_up: bool,
) -> TokenAmounts {
  let liq = BigDecimal::from_u128(liquidity).unwrap();
  let lower  = BigDecimal::from_u128(sqrt_price_x64_lower).unwrap();
  let upper = BigDecimal::from_u128(sqrt_price_x64_upper).unwrap();
  let current = BigDecimal::min(
    BigDecimal::max(
      BigDecimal::from_u128(sqrt_price_x64_current).unwrap(),
      lower.clone()),
    upper.clone()
  );

  let shift_64 = BigDecimal::from(BigInt::from_u128(1u128 << 64).unwrap());

  let token_a = liq.clone() * shift_64.clone() * (upper.clone() - current.clone()) / (current.clone() * upper.clone());
  let token_b = liq.clone() * (current.clone() - lower.clone()) / shift_64.clone();

  TokenAmounts {
    token_a: if !token_a.is_integer() && round_up { token_a.to_u64().unwrap() + 1 } else { token_a.to_u64().unwrap() },
    token_b: if !token_b.is_integer() && round_up { token_b.to_u64().unwrap() + 1 } else { token_b.to_u64().unwrap() },
  }
}

pub fn get_liquidity_from_token_a(
  sqrt_price_x64_0: u128,
  sqrt_price_x64_1: u128,
  amount: u64,
) -> u128 {
  let small_sqrt_price_x64 = BigInt::from_u128(u128::min(sqrt_price_x64_0, sqrt_price_x64_1)).unwrap();
  let large_sqrt_price_x64 = BigInt::from_u128(u128::max(sqrt_price_x64_0, sqrt_price_x64_1)).unwrap();
  if small_sqrt_price_x64 == large_sqrt_price_x64 {
    return 0u128;
  }

  let shift_64 = BigInt::from_u128(1u128 << 64).unwrap();
  let amount = BigInt::from_u64(amount).unwrap();
  let num = amount * (small_sqrt_price_x64.clone() * large_sqrt_price_x64.clone());
  let sub = large_sqrt_price_x64.checked_sub(&small_sqrt_price_x64).unwrap();
  let denom = shift_64.checked_mul(&sub).unwrap();
  num.checked_div(&denom).unwrap().to_u128().unwrap()
}

pub fn get_liquidity_from_token_b(
  sqrt_price_x64_0: u128,
  sqrt_price_x64_1: u128,
  amount: u64,
) -> u128 {
  let small_sqrt_price_x64 = BigInt::from_u128(u128::min(sqrt_price_x64_0, sqrt_price_x64_1)).unwrap();
  let large_sqrt_price_x64 = BigInt::from_u128(u128::max(sqrt_price_x64_0, sqrt_price_x64_1)).unwrap();
  if small_sqrt_price_x64 == large_sqrt_price_x64 {
    return 0u128;
  }

  let shift_64 = BigInt::from_u128(1u128 << 64).unwrap();
  let amount = BigInt::from_u64(amount).unwrap();
  let num = amount * shift_64;
  let denom = large_sqrt_price_x64.checked_sub(&small_sqrt_price_x64).unwrap();
  num.checked_div(&denom).unwrap().to_u128().unwrap()
}
